### DISCUSSION TITLE
is there anyway to rebuild only when db change using getStaticProps. "conditional serverside rendering"

### DISCUSSION THREAD

**Discussion by Sayed-Helmy** [vercel/next.js#33109] [Help] [answered]
### Describe the feature you'd like to request

is there any way to rebuild only when DB changes using getStaticProps? like if I'm watching my MongoDB and if there are some changes on this collection like a post request or delete or any change I wanna rebuild the page only on change is that possible? something like "conditional serverside rendering" only renders on DB change.

### Describe the solution you'd like

something like that .
`export const getStaticProps = async (ctx) => {
  let validate = false;
  const contact = await contactCollection();
  const watching = contact.watch();
  watching.on("change", () => {
    console.log("Change");
    validate = 1;
  });
  const data = await contact.find().toArray();
  const newData = data.map((data) => ({
    email: data.email,
    name: data.name,
    message: data.message,
  }));
  console.log(validate);
  return {
    props: {
      messages: newData,
    },
    revalidate: validate,
  };
};`

### Describe alternatives you've considered

if there are any other features in the next js now can do something like this it will be great for me ty "conditional serverside rendering"

**Comment by lachlanjc** [comment]
I've also wanted this feature! It'd be great if I could make a request to e.g. `/slug/_revalidate` or `/_next/revalidate/slug` when the data changes (e.g. in a CMS or database), then not have the site need to run the `revalidate` every x seconds. The API could look like `return { props: {}, revalidate: 'manual' }` or something. This would save at least tens of thousands of useless Vercel serverless function invocations per month across my projects, which right now are wasting electricity in order to be update-able quickly when data does change.

**Comment by ahmedaboutlaeb** [comment]
This feature should be available ASAP

**Comment by sicarius97** [comment]
@Darkmax512 What you and @lachlanjc probably want to look into is something like the SWR library (also worked on by the vercel/nextjs team). It has a lot more fine tuned options with data fetching and revalidation. It used the stale-while-revalidate method to cache data and reduces the amount of unnecessary requests that your app might be making when the data has not changed. You can also mutate the data in the cache without revalidating if you know what the updated data is going to look like already, which even further reduces the amount of calls to your api

SWR is also great because you can still statically generate your whole app and use it.

**Reply by Sayed-Helmy** [reply]
@sicarius97 yeah I'm using SWR hook in somewhere im my projects to make data fetching but SWR is a client-side data fetching I wanna make a conditional serverside rendering for any page I wanna if the props that page depends on change like when I watch a MongoDB for change that's the condition so I wanna next js to rebuild this page again only for this condition. ty

**Reply by sicarius97** [reply]
Sure but why not just make the page static and use data caching instead? It also seems as though you aren't really saving on any computation power by watching the db and only revalidating at an api endpoint if that data changes. Its really the same as just frequently pinging an api to refetch the data, you are just shifting the stress from your api to your db in that case

**Reply by sicarius97** [reply]
Also usually an SSR page in most all frameworks doesn't re-render 3xcept during a full client side page reload. You can't really force reload a page from the server-side once the client loads it, that would be unsafe for the client

**Comment by PsyGik** [comment]
You can follow this thread for updates: https://github.com/vercel/next.js/discussions/11552#discussioncomment-646963

**Comment by leerob** [comment, marked as answer]
I'm incredibly excited to share that Next.js 12.1 was just released, which includes **on-demand Incremental Static Regeneration**! We know many of you have been looking forward to this for a while, including myself, and I'm so happy it is here. 

- [Learn about Next.js 12.1](https://nextjs.org/blog/next-12-1)
- [Watch a demo of on-demand ISR](https://youtu.be/BGexHR1tuOA)
- [View another demo and give us your feedback](https://on-demand-isr.vercel.app/)
- [Read the documentation](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#on-demand-revalidation-beta)
